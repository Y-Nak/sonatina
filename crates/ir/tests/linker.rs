mod common;
use std::{
    env,
    path::{Path, PathBuf},
};

use common::parse_module;
use sonatina_ir::{ir_writer::ModuleWriter, module_linker::LinkedModule};

fn fixture_dir() -> PathBuf {
    let manifest_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
    let manifest_dir = Path::new(&manifest_dir);
    manifest_dir.join("tests/linker/fixtures")
}

macro_rules! test_ok {
    ($name:ident) => {
        #[test]
        fn $name() {
            let name = stringify!($name);
            let dir = fixture_dir().join("link_ok");

            let path_a = dir.join(format!("{name}_a.sntn"));
            let path_b = dir.join(format!("{name}_b.sntn"));
            let module_a = parse_module(path_a.to_str().unwrap());
            let module_b = parse_module(path_b.to_str().unwrap());

            let (linked, _) = LinkedModule::link(vec![module_a.module, module_b.module]).unwrap();
            let mut writer = ModuleWriter::new(linked.module());
            let module_text = writer.dump_string();

            let snap_file_path = dir.join(format!("{name}.snap"));
            snap_test!(module_text, snap_file_path.to_str().unwrap());
        }
    };
}

test_ok!(module);

macro_rules! test_error {
    ($name:ident) => {
        #[test]
        fn $name() {
            let name = stringify!($name);
            let dir = fixture_dir().join("link_errors");

            let path_a = dir.join(format!("{name}_a.sntn"));
            let path_b = dir.join(format!("{name}_b.sntn"));
            let module_a = parse_module(path_a.to_str().unwrap());
            let module_b = parse_module(path_b.to_str().unwrap());

            let linked = LinkedModule::link(vec![module_a.module, module_b.module]);
            assert!(linked.is_err());
        }
    };
}

test_error!(struct_error);
test_error!(gv_error);
test_error!(func_error);
test_error!(sig_error);
