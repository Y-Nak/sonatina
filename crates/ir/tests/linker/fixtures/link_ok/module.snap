---
source: crates/ir/tests/linker.rs
input_file: tests/linker/fixtures/link_ok/module.snap
---
target = evm-ethereum-cancun

type @foo = {i8, i16, *i64};
type @bar = {i8, [i8; 31]};

global public const i256 $ZERO = 0;
global external const i256 $ONE;

declare func external %f_outer(i8) -> i8;

func public %f_b(v0.i8, v1.i8) -> i8 {
    block0:
        v2.i8 = call %f_a v0;
        v3.i8 = add v0 v2;
        return;
}

func public %f_a(v0.i8) -> i8 {
    block0:
        return v0;
}

func private %types_a(v0.*i8, v1.[i8; 2], v2.[*i8; 2], v3.[[i8; 2]; 2], v4.@foo, v5.*@bar) -> i8 {
    block0:
        v6.i8 = mload v0 i8;
        v7.i8 = call %f_b v6 1.i8;
        call %f_outer;
        return v6;
}

func private %types_b(v0.*i8, v1.[i8; 2], v2.[*i8; 2], v3.[[i8; 2]; 2], v4.@bar, v5.*@foo) {
    block0:
        return;
}
